src/VintageBookshelf.Api/Controllers/BookshelvesController.cs

```diff
@@ -0,0 +1,103 @@
+using System.Threading.Tasks;
+using AutoMapper;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using VintageBookshelf.Api.Dtos;
+using VintageBookshelf.Domain.Interfaces;
+using VintageBookshelf.Domain.Models;
+using VintageBookshelf.Domain.Notifications;
+using VintageBookshelf.Domain.Services;
+
+namespace VintageBookshelf.Api.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/bookshelves")]
+    public class BookshelvesController : MainController
+    {
+        private readonly IBookshelfRepository _bookshelfRepository;
+        private readonly IBookshelfService _bookshelfService;
+        private readonly IMapper _mapper;
+
+        public BookshelvesController(IBookshelfRepository bookshelfRepository,
+                                     IBookshelfService bookshelfService,
+                                     IMapper mapper,
+                                     INotifier notifier) : base(notifier)
+        {
+            _bookshelfRepository = bookshelfRepository;
+            _bookshelfService = bookshelfService;
+            _mapper = mapper;
+        }
+        
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<ActionResult<IEnumerable<BookshelfDto>>> GetAll()
+        {
+            var bookshelves = _mapper.Map<IEnumerable<BookshelfDto>>(await _bookshelfRepository.GetAllWithProducts());
+            return CustomResponse(bookshelves);
+        }
+        
+        [HttpGet("{id:long}")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(BookshelfDto))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<ActionResult<BookshelfDto>> GetById(long id)
+        {
+            var bookshelf = _mapper.Map<BookshelfDto>(await _bookshelfRepository.GetBookshelfWithProducts(id));
+            if (bookshelf is null)
+            {
+                return NotFound();
+            }
+            return Ok(bookshelf);
+        }
+        
+        [HttpPost]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(BookshelfDto))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<BookshelfDto>> Add([FromBody] BookshelfDto bookshelfDto)
+        {
+            if (!ModelState.IsValid)
+            {
+                return CustomResponse(ModelState);
+            }
+            await _bookshelfService.Add(_mapper.Map<Bookshelf>(bookshelfDto));
+            return CustomResponse(bookshelfDto);
+        }
+        
+        [HttpPut("{id:long}")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(BookshelfDto))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<BookshelfDto>> Update(long id, [FromBody] BookshelfDto bookshelfDto)
+        {
+            if (id != bookshelfDto.Id)
+            {
```

Comment:
Considering the Id in the route, I would just remove the Id from BooksheldDto.

Reply:
Yes, it seems to just be redundant. Another nice catch!